app/routes/app._index.jsx

```diff
@@ -20,107 +19,99 @@ import { DiamondAlertMajor, ImageMajor } from "@shopify/polaris-icons";
-  const QRCodes = await getQRCodes(session.shop, admin.graphql);
+  const qrCodes = await getQRCodes(session.shop, admin.graphql);
 
   return json({
-    QRCodes,
+    qrCodes,
   });
 }
 // [END loader]
 
-export default function Index() {
-  const { QRCodes } = useLoaderData();
-  const navigate = useNavigate();
+// [START empty]
+const EmptyQRCodeState = ({ onAction }) => (
+  <EmptyState
+    heading="Create unique QR codes for your product"
+    action={{
+      content: "Create QR code",
+      onAction,
+    }}
+    image="https://cdn.shopify.com/s/files/1/0262/4071/2726/files/emptystate-files.png"
+  >
+    <p>Allow customers to scan codes and buy products using their phones.</p>
+  </EmptyState>
+);
+// [END empty]
 
-  function truncate(str) {
-    if (!str) return;
-    const n = 25;
-    return str.length > n ? str.substr(0, n - 1) + "…" : str;
-  }
+function truncate(str, { length = 25 } = {}) {
+  str = str || "";
+  if (str.length <= length) return str;
+  return str.slice(0, length) + "…";
+}
 
-  // [START empty]
-  const emptyMarkup = QRCodes.length ? null : (
-    <EmptyState
-      heading="Create unique QR codes for your product"
-      action={{
-        content: "Create QR code",
-        onAction: () => navigate("qrcodes/new"),
-      }}
-      image="https://cdn.shopify.com/s/files/1/0262/4071/2726/files/emptystate-files.png"
-    >
-      <p>Allow customers to scan codes and buy products using their phones.</p>
-    </EmptyState>
-  );
-  // [END empty]
+// [START table]
+const QRTable = ({ qrCodes }) => (
+  <IndexTable
+    resourceName={{
+      singular: "QR code",
+      plural: "QR codes",
+    }}
+    itemCount={qrCodes.length}
+    headings={[
+      { title: "Thumbnail", hidden: true },
+      { title: "Title" },
+      { title: "Product" },
+      { title: "Date created" },
+      { title: "Scans" },
+    ]}
+    selectable={false}
+  >
+    {qrCodes.map((qrCode, i) => (
+      <QRTableRow key={i} qrCode={qrCode} />
+    ))}
+  </IndexTable>
+);
+// [END table]
 
-  // [START table]
-  const qrCodesMarkup = QRCodes.length ? (
-    <IndexTable
-      resourceName={{
-        singular: "QR code",
-        plural: "QR codes",
-      }}
-      itemCount={QRCodes.length}
-      headings={[
-        { title: "Thumbnail", hidden: true },
-        { title: "Title" },
-        { title: "Product" },
-        { title: "Date created" },
-        { title: "Scans" },
-      ]}
-      selectable={false}
-    >
-      {/* [END table] */}
-      {QRCodes.map(
-        ({
-          id,
-          title,
-          productImage,
-          productTitle,
-          productDeleted,
-          createdAt,
-          scans,
-        }) => {
-          // [START row]
-          return (
-            <IndexTable.Row id={id} key={id} position={id}>
-              <IndexTable.Cell>
-                <Thumbnail
-                  source={productImage || ImageMajor}
-                  alt={"product image or placeholder"}
```

Comment:
There was a few bits of weird JSX like this, where `{}` was used for no reason.

I also fixed up the alt text to be more useful.